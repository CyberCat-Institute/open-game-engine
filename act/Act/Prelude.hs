{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE NoFieldSelectors #-}

module Act.Prelude (Word256, EthTransaction (..), Transaction (..), AbiType (..), AbiValue (..)) where

import Data.DoubleWord (Word256)
import Data.Text
import EVM.ABI (AbiType (..), AbiValue (..))
import EVM.Types (Addr, W256)

data Transaction = Transaction
  { contract :: String,
    method :: String,
    arguments :: [AbiValue]
  }
  deriving (Eq, Show, Ord)

data EthTransaction = EthTransaction
  { contract :: Addr,
    method :: Text,
    arguments :: [AbiValue],
    ethAmt :: W256,
    gas :: W256
  }
  deriving (Eq, Show, Ord)
