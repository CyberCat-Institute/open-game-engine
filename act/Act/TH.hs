{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TupleSections #-}

module Act.TH (act2OG, module Act.Prelude) where

import Act.Prelude
import Act.TH.Dispatch
import Act.TH.Extractor
import Act.TH.State
import Act.Utils
import CLI
import Data.List
import Data.Validation
import Error
import GHC.IO.Unsafe
import Language.Haskell.TH.Syntax as TH
import Syntax.Annotated

-- Convert from a an act filepath to a list of top-level declaration
-- state is converted into a record
-- invariants are converted into if-statements
-- We make one data declaration with one constructor for each method
-- which we then use for dispatching the correct functionality in the
-- contract.
-- a single top-level function is created taking in argument
-- the method dispatcher
act2OG :: String -> Q [Dec]
act2OG filename = do
  let file :: String = unsafePerformIO $ readFile filename
  let compiled :: Error String [Claim] =
        (compile $ file)
  case compiled of
    Failure err ->
      reportError (extractError err)
        >> pure []
    -- A parsed Act file is a list of claims
    Success val -> do
      contractFunction <- generateContractFunction val
      pure
        ( stateDec4Claim val
            ++ contractFunction
        )
  where
    extractError :: NonEmpty (Pn, String) -> String
    extractError err = ""

arr x y = AppT (AppT ArrowT x) y

-- generate a top-level function that will define the contract
-- it's type will always be the same:
-- (ContractState, ContractMethod) -> ContractState
-- This is then going to be instanciated as an open game using `fromFunctions`
generateContractFunction :: [Claim] -> Q [Dec]
generateContractFunction claims = do
  let (fnName, behaviors) = extractBehaviors claims
  let fnName' = (mkName (uncapitalise fnName ++ "Contract"))
  let methodInfo = getUpdates4Method behaviors
  methods <- traverse (\(a, b, c) -> (a,) <$> mapMethod2TH c b) methodInfo
  extractMethods <- generateExtractMethods (fmap (\(a, b, c) -> (a, c)) methodInfo)
  contractFn <- generateContractDecl fnName' methods extractMethods
  pure (contractFn)

contractState :: TH.Name
contractState = mkName "contractState"

contractArgs :: TH.Name
contractArgs = mkName "args"

-- This builds the function signature for the contract, each contract generates one function
-- that recieves transactions, each contract function has it's own internal "method dispatched"
-- which matches on the method of the transaction, extracts the arguments and calls the body with
-- the arguments in scope
-- Each contract is of this shape:
-- ```
-- contractName :: ContractState -> Transaction -> ContractState
-- contractName st transaction = case method transaction of
--     "m1" -> let (arg1, arg2) <- extractm1 (args transaction) in b1
--     "m2" -> let (arg1, arg2, arg3) <- extractm2 (args transaction) in b2
--     "m3" -> let (arg1) <- extractm3 (args transaction) in b3
-- ```
generateContractDecl :: Name -> [(String, TH.Exp)] -> [Dec] -> Q [Dec]
generateContractDecl fnName clauses extractMethods = do
  let method = mkName "method"
  crashMessage <- [|error ("unexpected method, got '" ++ $(return (VarE method)) ++ "'\\nexpected one of : " ++ $(return clauseLit))|]
  transactionPattern <- [p|Act.Prelude.Transaction _ $(return (VarP method)) $(return (VarP contractArgs))|]
  pure
    [ SigD
        fnName
        ((ConT (mkName "AmmState")) `arr` (ConT (mkName "Transaction") `arr` (ConT (mkName "AmmState")))),
      FunD
        fnName
        [ Clause
            [VarP contractState, transactionPattern]
            ( NormalB
                ( CaseE
                    (VarE method)
                    (matchClauses ++ [matchE WildP crashMessage])
                )
            )
            extractMethods
        ]
    ]
  where
    -- A string of all the methods, comma-separated. Used for printing errors.
    clauseLit :: TH.Exp
    clauseLit = LitE (StringL (intercalate ", " (fmap fst clauses)))

    -- This is the list of pattern to dispatch the transaction to the correct function body for the contract
    -- ```
    -- case (method transaction) of
    --   "call1" -> body1 \
    --   "call2" -> body2  |-> This bit
    --   "call3" -> body3 /
    -- ```
    matchClauses :: [Match]
    matchClauses = fmap (\(pat, exp) -> matchE (LitP (StringL pat)) exp) clauses
    matchE p b = Match p (NormalB b) []

undefinedE = VarE (mkName "undefined")

-- Extract the program associated with each method
-- Each program is a list of rewrite, the method is kept around as a string
-- The act compiler has two version of each method one for the "reverting"
-- behaviour of the method and one for the non-reverting behaviour (we know which
-- mode we're in if we check the `_mode` field and see if its `Pass`, the reverting
-- behavious is called `Fail`).
-- For Open Games we are only interested in the non-revering behaviour.
-- Transaction reversion will be handled at a later date if necessary. For now,
-- we don't have any obvious use-case for it.
getUpdates4Method :: [Behaviour] -> [(String, [Rewrite], Interface)]
getUpdates4Method behaviors =
  fmap (\x -> (_name x, _stateUpdates x, _interface x)) $ filter (\b -> _mode b == Pass) behaviors

-- Return the pattern for bringing the arguments into scope from the argument extractor
-- functions
-- If the method has no argument, the pattern is a unit
-- If the method has one argument the pattern is just a variable name
-- If the method has multiple arguments, the pattern is a tuple, one per argument
bindVariables :: [Decl] -> TH.Pat
bindVariables [] = VarP '()
bindVariables [Decl _ name] = VarP (mkName name)
bindVariables xs = TupP (fmap (VarP . mkName . getDeclId) xs)

-- A series of rewrites of the state is translated as the composition of the rewrites applied to the state
-- each rewrite is written as a single `State -> State` update function as per the implementation of `rewriteOne`
-- each constant is also written as a single `State -> State` update function
mapMethod2TH :: Interface -> [Rewrite] -> Q TH.Exp
mapMethod2TH (Interface methodName args) actExp = do
  -- body <- [| ($(foldl1 (\e1 e2 -> [|$e2 . $e1|]) (fmap rewriteOne actExp))) $(return (VarE contractState)) |]
  body <- RecUpdE (VarE contractState) <$> traverse rewriteOne actExp
  let extractor = AppE (VarE (argumentExtractorName methodName)) (VarE contractArgs)
  pure $ LetE [ValD (bindVariables args) (NormalB extractor) []] body
  where
    rewriteOne :: Rewrite -> Q (Name, TH.Exp)
    rewriteOne (Constant loc) = pure (mkName "unimplemented", VarE (mkName "undefined"))
    rewriteOne (Rewrite (Update ty (Item _ nm varName _) newValue)) =
      (mkName varName,) <$> mapExp newValue

-- The rest of this file is for debugging purposes

-- display :: String
-- display = case compiled of
--   Success s -> unlines (fmap ((++ "\n") . printClaim) s)

printBehaviour :: Behaviour -> String
printBehaviour b =
  "name: "
    ++ _name b
    ++ "\n"
    ++ "mode: "
    ++ show (_mode b)
    ++ "\n"
    ++ "contract: "
    ++ show (_contract b)
    ++ "\n"
    ++ "interface: "
    ++ show (_interface b)
    ++ "\n"
    ++ "preconditions:\n"
    ++ unlines (fmap (("  - " ++) . show) (_preconditions b))
    ++ "postConditions:\n"
    ++ unlines (fmap (("  - " ++) . show) (_postconditions b))
    ++ "\nstate updates:\n"
    ++ unlines (fmap (("  - " ++) . show) (_stateUpdates b))

printClaim :: Claim -> String
printClaim (C constructor) = "constructor: " ++ show constructor
printClaim (B b) = "behaviour:\n" ++ printBehaviour b
printClaim (I i) = "invariant: " ++ show i
printClaim (S s) = "Storage: " ++ show s

printConstructor :: Constructor -> String
printConstructor
  ( Constructor
      name
      mode
      interface
      preconditions
      postconditions
      initialStorage
      stateupdates
    ) =
    "constructor: " ++ name
