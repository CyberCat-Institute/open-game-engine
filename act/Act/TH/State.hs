{-# LANGUAGE TemplateHaskell #-}

module Act.TH.State where

import Act.Utils
import Data.Bifunctor
import qualified Data.Map as M
import Language.Haskell.TH.Syntax
import Syntax.Untyped (Creates(..), SlotType)
import Syntax.Annotated (Id)

-- Generate a type for the global state of the contract
stateDec4Claim :: Creates -> [Dec]
stateDec4Claim (Creates assign) = [createDataDeclaration undefined]
--   where
--     convertStorage :: Claim -> [Dec]
--     convertStorage (C c) = []
--     convertStorage (B b) = []
--     convertStorage (I i) = []
--     convertStorage (S s) = fmap (createDataDeclaration . second M.toList) (M.toList s)

createDataDeclaration :: (Id, [(Id, SlotType)]) -> Dec
createDataDeclaration (storeName, types) =
  DataD
    [] -- no constraints
    (storeTypeName) -- the name is the name from the store
    [] -- no type variables
    Nothing -- no kind signature, not a GADT
    [RecC storeTypeName (fmap (\(nm, ty) -> (mkName nm, defaultBang, mapEVMTypes ty)) types)]
    [DerivClause Nothing [ConT ''Show, ConT ''Eq]]
  where
    storeTypeName :: Name
    storeTypeName = mkName (capitalise (storeName ++ "State"))
