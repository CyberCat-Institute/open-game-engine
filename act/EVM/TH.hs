{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TupleSections #-}

module EVM.TH where

import Act.Prelude (EthTransaction (..))
import Control.Monad.Trans.State.Strict (State, put)
import Data.ByteString (ByteString)
import Data.Map as Map
import Data.Text (Text, unpack)
import Data.Text.IO (readFile)
import qualified Data.Tree.Zipper as Zipper
import Data.Vector as Vector (fromList)
import EVM (blankState, initialContract)
import EVM.ABI
import EVM.FeeSchedule
import EVM.Solidity (solcRuntime)
import EVM.Types
import GHC.IO.Unsafe
import Language.Haskell.TH.Syntax as TH
import Prelude hiding (FilePath, readFile)

-- put this in sttate.callData
-- run it to execute the transaction
-- put more for subsequent calls
-- run more for more results
makeCallData :: EthTransaction -> ByteString
makeCallData (EthTransaction _ method args _ _) =
  abiMethod method (AbiTuple (Vector.fromList args))

sendTransaction :: Contract -> EthTransaction -> VM -> VM
sendTransaction cts tx st = st { frames = [txFrame] }
  where
  txFrame :: Frame
  txFrame = Frame frameCtx frameState

  frameCtx :: FrameContext
  frameCtx = CallContext
      { target        = tx.contract
      , context       = 0
      , offset        = 0
      , size          = 0
      , codehash      = cts.codehash
      , abi           = Nothing
      , calldata      = ConcreteBuf (makeCallData tx)
      , callreversion = (undefined, undefined)
      , subState      = SubState mempty mempty mempty mempty mempty
      }

  frameState :: FrameState
  frameState = FrameState
      { contract     = tx.contract
      , codeContract = undefined
      , code         = cts.contractcode
      , pc           = 0
      , stack        = mempty
      , memory       = undefined
      , memorySize   = 0
      , calldata     = ConcreteBuf (makeCallData tx)
      , callvalue    = Lit 0
      , caller       = Lit 0
      , gas          = tx.gas
      , returndata   = undefined
      , static       = undefined
      }

emptyVM :: [(Addr, ByteString)] -> VM
emptyVM contracts =
  VM
    { result = Nothing,
      state = blankState,
      frames = [],
      env = envForContracts contracts,
      block = emptyBlock,
      tx = emptyTransaction,
      logs = [],
      traces = Zipper.fromForest mempty,
      cache = Cache mempty mempty mempty,
      burned = 0,
      iterations = mempty,
      constraints = [],
      keccakEqs = [],
      allowFFI = True,
      overrideCaller = Nothing
    }
  where
    -- question: Is that a reasonable empty first block?
    emptyBlock :: Block
    emptyBlock =
      Block
        { coinbase = 0,
          timestamp = Lit 0,
          number = 0,
          prevRandao = 0,
          maxCodeSize = 0,
          gaslimit = 0,
          baseFee = 0,
          schedule = berlin -- specifically this, what is it suppsoed to be?
        }
    emptyTransaction :: TxState
    emptyTransaction =
      TxState
        { gasprice = 0,
          gaslimit = 0,
          priorityFee = 0,
          origin = 0,
          toAddr = 0,
          value = Lit 0,
          substate = emptySubState,
          isCreate = True,
          txReversion = mempty
        }
    emptySubState :: SubState
    emptySubState =
      SubState
        { selfdestructs = [],
          touchedAccounts = [],
          accessedAddresses = mempty,
          accessedStorageKeys = mempty,
          refunds = []
        }

    envForContracts :: [(Addr, ByteString)] -> Env
    envForContracts contracts =
      Env
        { contracts = Map.fromList (fmap (fmap bytecodeToContract) contracts),
          chainId = 0,
          storage = EmptyStore,
          origStorage = mempty,
          sha3Crack = mempty
        }

    bytecodeToContract :: ByteString -> Contract
    bytecodeToContract = initialContract . RuntimeCode . ConcreteRuntimeCode

-- setup a new VM state from the list of contracts we are using
loadIntoVM :: [(Addr, ByteString)] -> State VM ()
loadIntoVM contracts = put (emptyVM contracts)

-- import a list of contracts as an open game
-- - first we read off all the files and translate them into solidity bytecode
-- - Then we associate each contract to a contract name which
loadEVM :: [(Text, Text)] -> IO (State VM ())
loadEVM contracts = do
  files :: [(Text, Text)] <- traverse (\(name, filename) -> (name,) <$> readFile (unpack filename)) (contracts)
  contracts :: [ByteString] <-
    traverse
      ( \(nm, body) -> do
          Just bytecode <- solcRuntime nm body
          pure bytecode
      )
      files
  let bytecodeMap :: [(Addr, ByteString)] = zip [0 ..] contracts
  let newVM = loadIntoVM bytecodeMap
  pure newVM

loadContracts :: [(Text, Text)] -> State VM ()
loadContracts arg = unsafePerformIO $ loadEVM arg

compileTimeLoad :: [(Text, Text)] -> Q [Dec]
compileTimeLoad = undefined
