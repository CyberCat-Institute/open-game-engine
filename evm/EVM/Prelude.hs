{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE NoFieldSelectors #-}

module EVM.Prelude (Word256, EthTransaction (..), AbiType (..), AbiValue (..)) where

import Data.DoubleWord (Word256)
import Data.Text
import EVM.ABI (AbiType (..), AbiValue (..))
import EVM.Types (Addr, W256)
import GHC.Word

data EthTransaction = EthTransaction
  { contract :: Addr,
    caller :: Addr,
    method :: Text,
    arguments :: [AbiValue],
    ethAmt :: W256,
    gas :: Word64
  }
  deriving (Eq, Show, Ord)
