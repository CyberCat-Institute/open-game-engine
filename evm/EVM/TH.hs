{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TupleSections #-}

module EVM.TH where

import Control.Monad.Trans.State.Strict
import Control.Monad.Trans.State.Strict (State, put)
import Control.Monad.ST
import Data.ByteString (ByteString)
import Data.Map as Map
import Data.Text (Text, pack, unpack, intercalate, toLower)
import Data.Text.IO (readFile)
import qualified Data.Tree.Zipper as Zipper
import Data.Vector as Vector (fromList)
import Debug.Trace
import EVM (blankState, initialContract, loadContract, resetState, exec1)
import EVM.Exec (exec, run)
import EVM.Expr
import EVM.FeeSchedule
import EVM.Fetch
import EVM.Prelude
import EVM.Solidity (solcRuntime, readStdJSON, solidity', Contracts(..), SolcContract(..), Method(..))
import EVM.Stepper
import EVM.Transaction (initTx)
import EVM.Types
import GHC.IO.Unsafe
import GHC.ST
import Language.Haskell.TH.Syntax as TH
import Optics.Core
import Optics.State
import Optics.State.Operators
import Prelude hiding (FilePath, readFile)

-- put this in sttate.callData
-- run it to execute the transaction
-- put more for subsequent calls
-- run more for more results
makeCallData :: EthTransaction -> Expr Buf
makeCallData (EthTransaction _ caller method args _ _) =
  ConcreteBuf $ abiMethod method (AbiTuple (Vector.fromList args))

makeTxCall :: EthTransaction -> EVM s ()
makeTxCall tx@(EthTransaction addr caller meth args amt gas) = do
  resetState
  assign (#tx % #isCreate) False
  modify (execState (loadContract addr))
  assign (#state % #callvalue) (Lit amt)
  assign (#state % #calldata) (makeCallData tx)
  assign (#state % #caller) (caller)
  assign (#state % #gas) gas
  modify initTx

loadIntoVM :: [(Expr EAddr, ByteString)] -> ST s (VM s)
loadIntoVM contracts = do
  blankSt <- blankState
  pure $
    VM
      { result = Nothing,
        state = blankSt,
        frames = [],
        env = envForContracts contracts,
        block = emptyBlock,
        tx = emptyTransaction,
        logs = [],
        traces = Zipper.fromForest mempty,
        cache = Cache mempty mempty,
        burned = 0,
        iterations = mempty,
        constraints = [],
        keccakEqs = [],
        config =
          RuntimeConfig
            True
            Nothing
            EmptyBase
      }
  where
    -- question: Is that a reasonable empty first block?
    emptyBlock :: Block
    emptyBlock =
      Block
        { coinbase = LitAddr 0,
          timestamp = Lit 0,
          number = 0,
          prevRandao = 0,
          maxCodeSize = 0,
          gaslimit = 0,
          baseFee = 0,
          schedule = feeSchedule
        }
    emptyTransaction :: TxState
    emptyTransaction =
      TxState
        { gasprice = 0,
          gaslimit = 0,
          priorityFee = 0,
          origin = LitAddr 0,
          toAddr = LitAddr 0,
          value = Lit 0,
          substate = emptySubState,
          isCreate = True,
          txReversion = mempty
        }
    emptySubState :: SubState
    emptySubState =
      SubState
        { selfdestructs = [],
          touchedAccounts = [],
          accessedAddresses = mempty,
          accessedStorageKeys = mempty,
          refunds = []
        }

    envForContracts :: [(Expr EAddr, ByteString)] -> Env
    envForContracts contracts =
      Env
        { contracts = Map.fromList (fmap (fmap bytecodeToContract) contracts),
          chainId = 0
          -- storage = EmptyStore,
          -- origStorage = mempty
        }

    bytecodeToContract :: ByteString -> Contract
    bytecodeToContract = initialContract . RuntimeCode . ConcreteRuntimeCode


int :: Int -> Exp
int = LitE . IntegerL . toInteger

constructorExprForType :: AbiType -> Name -> Exp
constructorExprForType (AbiUIntType w) = ((ConE (mkName "AbiUInt") `AppE` int w) `AppE`) . VarE
constructorExprForType (AbiIntType w) = ((ConE (mkName "AbiInt") `AppE` int w) `AppE`) . VarE
constructorExprForType (AbiAddressType) = (ConE (mkName "AbiAddress") `AppE`) . VarE
constructorExprForType (AbiBoolType) = (ConE (mkName "AbiBool") `AppE`) . VarE
constructorExprForType (AbiBytesType w) = ((ConE (mkName "AbiBytes") `AppE` int w) `AppE`) . VarE
constructorExprForType (AbiBytesDynamicType) = (ConE (mkName "AbiBytesDynamic") `AppE`) . VarE
constructorExprForType (AbiStringType) = (ConE (mkName "AbiString") `AppE`) . VarE
constructorExprForType (AbiArrayDynamicType ty) = error "arrays unsupported"
constructorExprForType (AbiArrayType size ty) = error "arrays unsuppported"
constructorExprForType (AbiTupleType types) = error "tuples unsupported" -- ConE (mkName "AbiTuple") [] [VarP (mkName name)]
constructorExprForType (AbiFunctionType) = error "functions unsupported"

data ContractInfo = ContractInfo
    { file :: Text
    , name :: Text
    , boundName :: Text
    }

-- loadContracts :: [ContractInfo] -> ST s (VM s)
-- loadContracts arg = unsafePerformIO $ loadEVM arg

-- import a list of contracts as an open game
-- - first we read off all the files and translate them into solidity bytecode
-- - Then we associate each contract to a contract name which
-- loadEVM :: [ContractInfo] -> IO (ST s (VM s))
-- loadEVM contracts = do
--   files :: [ContractInfo] <-
--       traverse (\(ContractInfo filename name bound) ->
--                   ContractInfo <$> readFile (unpack filename)
--                                <*> pure name
--                                <*> pure bound) contracts
--   contracts :: [(Text, Text, ByteString)] <-
--     traverse
--       ( \(ContractInfo fileBody nm bound) -> do
--           bytecode <- solcRuntime nm body
--           maybe (error "solc failed") (pure . (nm, bound,)) bytecode
--       )
--       files
--   let bytecodeMap :: [(Expr EAddr, ByteString)] = zip (fmap LitAddr [0xabcd ..]) contracts
--   let newVM = loadIntoVM bytecodeMap
--   pure newVM

pat = VarP . mkName

generateTxFactory :: Method -> Integer -> Text -> Q Dec
generateTxFactory (Method _ args name sig _) addr boundName = do
  runIO $ print ("arguments for method " <> name <> ":" <> pack (show args))
  let signatureString :: Q Exp = pure $ LitE $ StringL $ unpack sig
  let argExp :: Q Exp = pure $ ListE $ fmap (\(nm, ty) -> constructorExprForType ty (mkName $ unpack nm)) args
  let patterns :: [Pat] = fmap (VarP . mkName . unpack . fst) args
  let contractAddress :: Q Exp = pure $ AppE (ConE (mkName "LitAddr")) (LitE (IntegerL addr))
  body <- [e| EthTransaction
                $(contractAddress)
                src
                $(signatureString)
                $(argExp)
                amt gas|]
  pure $ FunD (mkName (unpack (toLower boundName <> "_" <> name)))
              [Clause ([pat "src", pat "amt", pat "gas"]
                      ++ patterns)
                      (NormalB body) []]

instance Lift (Addr) where
  lift (Addr word) = let v = toInteger word in [e| fromInteger v |]

instance Lift (Expr 'EAddr) where
  lift (LitAddr a) = [e| LitAddr (fromInteger a) |]

loadAll :: [ContractInfo] -> Q [Dec]
loadAll contracts = do
  cs <- runIO $ traverse loadSolcInfo contracts
  let indexed = zip [0x1000 ..] cs
  methods <- traverse (\(ix, (_, bn, _, m)) -> traverse (\m' -> generateTxFactory m' ix bn) m) indexed
  let allMethods = concat methods
  let contractMap = fmap (\(ix, (_, _, b, _)) -> (LitAddr (fromInteger ix), b)) indexed
  init <- [d|
      initial :: ST s (VM s)
      initial = loadIntoVM contractMap
       |]
  pure (init ++ allMethods)

loadSolcInfo :: ContractInfo -> IO (Text, Text, ByteString, [Method])
loadSolcInfo (ContractInfo contractFilename contractName boundName) = do
  file <- readFile (unpack contractFilename)
  solRes <- solidity' file
  let (bytecode, methods) = getSolcResults solRes
  pure (contractName, boundName, bytecode, methods)
  where

  getSolcResults :: (Text, Text) -> (ByteString, [Method])
  getSolcResults (json, path) = case readStdJSON json of
    Nothing -> error ("could not read json file: " <> show json)
    Just (Contracts sol, a, b) ->
        case Map.lookup (path <> ":" <> contractName) sol of
          Nothing -> error "failed looking up contract"
          Just contract -> let methods = Map.elems $ contract.abiMap
                            in (contract.runtimeCode, methods)

run' :: EVM s (VM s)
run' = do
  vm <- get
  case vm.result of
    Nothing -> exec1 >> run'
    Just (HandleEffect (Query (PleaseAskSMT (Lit c) _ cont))) -> cont (Case (c > 0)) >> run'
    Just _ -> pure vm

-- send and run a transaction on the EVM state
sendAndRun' :: EthTransaction -> EVM RealWorld (VM RealWorld)
sendAndRun' tx = do
  EVM.TH.makeTxCall tx
  vm <- run'
  traceM (show vm.result)
  pure vm

-- exectute the EVM state in IO
sendAndRun ::
  EthTransaction ->
  VM RealWorld ->
  EVM RealWorld (VM RealWorld)
sendAndRun tx st = do
  put st
  sendAndRun' tx
  get



-- TODO: use foundry
-- thatOneMethod =
--   let st = loadContracts [ContractInfo "solidity/Simple.sol" "Neg" "test"]
--       ourTransaction =
--         EthTransaction
--           (LitAddr 0xabcd)
--           (LitAddr 0x1234)
--           "negate(int256)"
--           [AbiInt 256 3]
--           100000000
--           100000000
--       steps = do
--         evm (makeTxCall ourTransaction)
--         runFully
--    in interpret (zero 0 (Just 0)) undefined steps
