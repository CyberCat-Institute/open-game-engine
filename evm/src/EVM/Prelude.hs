{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE NoFieldSelectors #-}

module EVM.Prelude (Word256, EthTransaction (..), AbiType (..), AbiValue (..), abiMethod) where

import Data.DoubleWord (Word256)
import Data.Text
import EVM.ABI (AbiType (..), AbiValue (..), abiMethod)
import EVM.Types (Addr, EType (..), Expr (..), W256)
import GHC.Word

data EthTransaction = EthTransaction
  { contract :: Expr EAddr,
    caller :: Expr EAddr,
    method :: Text,
    arguments :: [AbiValue],
    ethAmt :: W256,
    gas :: Word64
  }
  deriving (Eq, Show, Ord)
