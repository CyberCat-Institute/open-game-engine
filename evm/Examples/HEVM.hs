{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NumericUnderscores #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}

module Examples.HEVM where

import Control.Monad.ST
import Control.Monad.Trans.State.Strict
import Data.DoubleWord
import qualified Data.Map as Map
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as T
import Debug.Trace
import EVM (emptyContract, exec1)
import EVM.Exec
import EVM.Fetch (zero)
import EVM.Format
import EVM.Prelude
import EVM.Stepper (evm, interpret, runFully)
import EVM.TH
import EVM.Types
import GHC.Float
import OpenGames hiding (dependentDecision, fromFunctions, fromLens)
import OpenGames.Engine.Diagnostics
import OpenGames.Engine.HEVMGames
import OpenGames.Preprocessor hiding (Lit)
import Optics.Core (at, over, preview, set, view, (%), (%?), (&), (.~))

$(loadAll [ContractInfo "solidity/Withdraw.sol" "Piggybank" "store"])

deposit :: EthTransaction
deposit = store_deposit userContractAddress 1000 10_000_000

dummyTx :: Word256 -> EthTransaction
dummyTx = store_retrieve userContractAddress 0 100000000

userContractAddress = LitAddr 0x1234

withdrawContractAddress = LitAddr 0x1000

transactionList :: Word256 -> [EthTransaction]
transactionList max = [dummyTx n | n <- [1 .. max]]

playerManual globalState =
  [opengame|
  inputs   : ;
  feedback : ;
  :-------:

  operation : hevmDecision "AllPlayers" (transactionList 2) ;
  outputs   : transactions ;
  returns   : balance finalState userContractAddress;

  inputs    : transactions, globalState;
  feedback  : ;
  operation : fromLensM (uncurry sendAndRun) (const pure) ;
  outputs   : finalState ;
  returns   : ;

  :-------:
  outputs: ;
  returns : ;
|]

playerAutomatic =
  [opengame|
  inputs   : ;
  feedback : ;
  :-------:

  operation : hevmDecision "AllPlayers" (transactionList 2) ;
  outputs   : transactions ;
  returns   : balance finalState userContractAddress;

  inputs    : transactions;
  feedback  : ;
  operation : fromLensM sendAndRun' (const pure) ;
  outputs   : finalState ;
  returns   : ;

  :-------:
  outputs: ;
  returns : ;
|]

outcome = do
  i <- stToIO initial
  let newI = setupAddresses [(userContractAddress, Lit 1_000_000_000)] i
  newI <- interpret (zero 0 (Just 0)) newI (evm (makeTxCall deposit) >> runFully)
  let term :- Nil = evaluate (playerManual newI) ((pure (dummyTx 1)) :- Nil) void
  let t' = evalStateT term newI
  tevaluated <- stToIO t'
  generateOutput ([tevaluated] :- Nil)

outcomeAutomatic = do
  i <- stToIO initial
  putStrLn "initial contracts:"
  print $ getAllContracts i
  let newI = setupAddresses [(userContractAddress, Lit 1_000_000_000),
                             (LitAddr 0x1000, Lit 1000)] i
  putStrLn "setup contracts:"
  print $ getAllContracts newI
  newI <- interpret (zero 0 (Just 0)) newI (evm (makeTxCall deposit) >> runFully)
  let term :- Nil = evaluate (playerAutomatic) ((pure (dummyTx 1)) :- Nil) void
  let t' = evalStateT term newI
  putStrLn "end contracts:"
  print $ getAllContracts newI
  tevaluated <- stToIO t'
  generateOutput ([tevaluated] :- Nil)

testExec = do
  evm $ makeTxCall deposit
  -- evm $ makeTxCall (dummyTx 2)
  runFully

showVM :: VM s -> Text
showVM vm =
  T.unlines
    [ "Contracts:",
      indent 2 . T.unlines . Map.elems $ Map.mapWithKey (\a c -> T.pack (show a) <> " :\n  " <> showContract c) vm.env.contracts,
      -- , "Storage: " <> (formatExpr vm.env.storage)
      --
      "CallValue: " <> (formatExpr vm.state.callvalue),
      "Result: " <> (T.pack $ show vm.result)
    ]

showContract :: Contract -> Text
showContract c =
  T.unlines
    [ "balance: " <> (T.pack $ show c.balance)
    ]

interp = do
  i <- stToIO initial
  let newI = setupAddresses [(userContractAddress, Lit 1_000_000_000)] i
  newI <- interpret (zero 0 (Just 0)) newI (evm (makeTxCall deposit) >> runFully)
  let storage = preview (#env % #contracts % at withdrawContractAddress %? #storage) newI
  let orig = preview (#env % #contracts % at withdrawContractAddress %? #origStorage) newI
  traceM ("storage: " <> show storage)
  traceM ("origStorage: " <> show orig)
  T.putStrLn (showVM newI)
