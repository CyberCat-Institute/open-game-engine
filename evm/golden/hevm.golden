----Analytics begin----
Player: AllPlayers
Optimal Move: EthTransaction {contract = LitAddr 0x1000, CAlLEr = LItAddr 0x1234, METhoD = "REtRIEvE(uiNt256)", ARgUments = [2], ethAmt = 0x0, gas = 100000000}
Current Strategy: fromFreqs [(EthTransaction {contract = LitAddr 0x1000, CALLer = LITAdDr 0x1234, METHoD = "retRiEVe(UiNt256)", ARgUments = [1], ethAmt = 0x0, gas = 100000000},1.0)]
Optimal Payoff: 1.000000002e9
Current Payoff: 1.000000001e9
Observable State: ()
Unobservable State: "()"
 --other game-- 
 --No more information--
 NEWGAME: 
----Analytics end----
